src/dune_file_watcher/dune_file_watcher.ml

```diff
@@ -138,48 +152,70 @@ let command ~root =
         ; [ "--format"; "e:%e:%w%f" ]
         ; [ "-m" ]
         ]
-    , Inotify.parse_message
-    , Some Inotify.wait_for_watches_established )
-  | None -> (
+    , Inotifywait.parse_message
+    , Some Inotifywait.wait_for_watches_established )
+  | `Fswatch fswatch ->
     (* On all other platforms, try to use fswatch. fswatch's event filtering is
        not reliable (at least on Linux), so don't try to use it, instead act on
        all events. *)
-    match Bin.which ~path:(Env.path Env.initial) "fswatch" with
-    | Some fswatch ->
-      let excludes =
-        List.concat_map
-          (exclude_patterns @ List.map exclude_paths ~f:(fun p -> "/" ^ p))
-          ~f:(fun x -> [ "--exclude"; x ])
-      in
-      ( fswatch
-      , [ "-r"
-        ; root
-        ; (* If [inotify_special_path] is not passed here, then the [--exclude
-             _build] makes fswatch not descend into [_build], which means it
-             never even discovers that [inotify_special_path] exists. This is
-             despite the fact that [--include] appears before. *)
-          inotify_special_path
-        ; "--event"
-        ; "Created"
-        ; "--event"
-        ; "Updated"
-        ; "--event"
-        ; "Removed"
-        ]
-        @ [ "--include"; inotify_special_path ]
-        @ excludes
-      , (fun s -> Ok s)
-      , None )
-    | None ->
-      User_error.raise
-        [ Pp.text
-            (if Sys.linux then
-              "Please install inotifywait to enable watch mode. If inotifywait \
-               is unavailable, fswatch may also be used but will result in a \
-               worse experience."
-            else
-              "Please install fswatch to enable watch mode.")
-        ])
+    let excludes =
+      List.concat_map
+        (exclude_patterns @ List.map exclude_paths ~f:(fun p -> "/" ^ p))
+        ~f:(fun x -> [ "--exclude"; x ])
+    in
+    ( fswatch
+    , [ "-r"
+      ; root
+      ; (* If [inotify_special_path] is not passed here, then the [--exclude
+           _build] makes fswatch not descend into [_build], which means it never
+           even discovers that [inotify_special_path] exists. This is despite
+           the fact that [--include] appears before. *)
+        inotify_special_path
+      ; "--event"
+      ; "Created"
+      ; "--event"
+      ; "Updated"
+      ; "--event"
+      ; "Removed"
+      ]
+      @ [ "--include"; inotify_special_path ]
+      @ excludes
+    , (fun s -> Ok s)
+    , None )
+
+let select_watcher_backend ~use_inotify_lib =
+  let try_fswatch () =
+    Option.map
+      (Bin.which ~path:(Env.path Env.initial) "fswatch")
+      ~f:(fun fswatch -> `Fswatch fswatch)
+  in
+  let try_inotifywait () =
+    Option.map
+      (Bin.which ~path:(Env.path Env.initial) "inotifywait")
+      ~f:(fun inotifywait -> `Inotifywait inotifywait)
+  in
+  let error str = User_error.raise [ Pp.text str ] in
+  match Sys.linux with
```

Comment:
To keep things in sync in case we need to change the `#ifdef` in `inotify_stubs.c`, we could add `Inotify.available : bool` which would be implemented via a C function and would reflect whether we are using the real or dummy implementation of the stubs.

Reply:
I considered this (in fact I wrote the stubs), but then I thought that:
- The check based on `Sys.linux` is already pretty much the same thing
- The error raised by the stubs (`ENOSUPP`) should indicate clearly enough if something goes out of sync

so removed those stubs for the sake of simplicity.

Reply:
The commit is c5ccfd8755305481da333ea5a06efc1bca390ab2. Do you feel like it's worth keeping or we're happier without?

Reply:
I would personally keep it. Just to keep things tight, and a bit more explicit as well. For instance, so that you don't need to know that inotify is linux only when reading this code.

Reply:
Yeah, good point about not knowing upfront that inotify is linux-only.